Expensify/App
# Reauthenticate Pusher on Reconnect

**Fixed issues:** 

- Report view navigating to `/` after reauthenticating with expired `authToken`
- Pusher using old authToken when reconnecting and updating `authToken`
- Network request queue was not emptying all the way
- Network requests queued when authToken expired were never actually getting queued resulting in missing data
- Re-authentication bug prevented us from consistently re-authenticating with stored credentials

**Tests:**

The majority of these tests were performed with an `authToken` that expires every 5 seconds to create a ton of instability. I wanted to make sure that if our `authToken` expires while we step away that everything doesn't go to hell. So to create this situation you'll to make some auth changes on dev by using this diff and rebuilding server-e

```diff
diff --git a/auth/lib/AuthToken.cpp b/auth/lib/AuthToken.cpp
index 0e08652f2..a82240603 100644
--- a/auth/lib/AuthToken.cpp
+++ b/auth/lib/AuthToken.cpp
@@ -776,7 +776,7 @@ string AuthToken::create(int64_t accountID, int64_t partnerID, const bool isVali
     authToken["accountID"] = to_string(accountID);
     authToken["partnerID"] = to_string(partnerID);
     authToken["issued"] = to_string(STimeNow());
-    authToken["expires"] = to_string(STimeNow() + expires);
+    authToken["expires"] = to_string(STimeNow() + 5000000);
     authToken["type"] = typeStrings[type];

     if (requestingAccountID != 0) {
```

clear `localStorage`

**Testing that an authToken can expire while we are in the middle of leaving a comment and we send it like nothing happened**
1. Login with your 5 second `authToken`
1. Wait 5 seconds and leave a comment on a report
1. Dig through the network requests and notice that the initial API requests that failed eventually completed
1. Refresh and verify that the history is now permanent

**Testing that if we refresh the page we can reconnect to Pusher**

This was a weird one because Pusher would intermittently stop working so try to refresh the page many times and with and without reportID in the url

1. Wait 5 seconds
1. Refresh the page
1. Notice that in the debug logs say your `authToken` has expired
1. Use `getPusherInstance()` to inspect the private user channel and see that you are still subscribed

**Testing that when WiFi drops with an expired `authToken` we are still connected to pusher on re-connect**

1. Wait 5 seconds
1. Turn off wifi
1. Wait for Pusher to start throwing errors - 10-15 seconds
1. Turn wifi back on
1. Notice that in the debug logs say your `authToken` has expired
1. Use `getPusherInstance()` to inspect the private user channel and see that you are still subscribed
